### PR TITLE
[Messenger] Update External Messenger Implementation

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
Update the external messenger implementation to unify with internal api/architecture.

This diff is rather large. I'll explain the details below:

1. The concept of `parlai/messenger/tasks/<task>/run.py` is now obsolete. Options for tasks are specified in `tasks/<task>/config.yml`
2. As a consequence of using `configs` instead of `<task>/run.py`, `World`s now define the following static methods:
    - `generate_world` - returns an instantiated world
    - `assign_roles` - assigns roles to agents
    - `eligibility_function` - determines whether an agent is eligible for the task
3. To run a task, one runs the `parlai/messenger/core/run.py` script, with one argument, `--config-path`, a path to a `config.yml`
4. Worlds are now run via `futures` objects, rather than threads.
5. Worlds are now run via a `MessengerWorldRunner` object, which exposes two functions, `launch_overworld` and `launch_task_world`, which run overworlds and task worlds.
6. Onboarding is now shifted out of the `MessengerManager` and into the `WorldRunner`
7. The `MessengerManager` is now responsible for loading models.
8. Several functions have been added to `shared_utils`, primarily for parsing the config files and retrieving various functions from worlds.
9. All demo tasks have been updated to work with the appropriate new logic.

One big **TODO** that is absolutely necessary before merging is that I need to remove the `page_id` from the respective configs - I have left in there in case any of you wanted to test this out yourself using that page.

**Testing steps**
I tried running each demo world and confirmed that they all seem to work. Do we have any unit testing for messenger? Curious to see if that works

**Logs**
I'll include screenshots of me running the various demo worlds below:

_Chatbot_
Screenshot 2019-10-07 17 33 59

_QADataCollection_
Screenshot 2019-10-07 17 55 57

_OverworldDemos_
Screenshot 2019-10-07 18 19 32
Screenshot 2019-10-07 18 16 36
Screenshot 2019-10-07 18 16 06

### CODE REVIEW

File: parlai/messenger/core/messenger_manager.py [hunk 1]
```diff
@@ -156,22 +270,24 @@ def _expire_all_conversations(self):
...
+    def _get_unique_pool(self):
+        """Return unique pool.
+
+        Returns a filtered version of the agent pool where each agent is
         only listed a maximum of one time.
+
+        :return:
+            a dictionary mapping world_types to agent pools
         """
         # TODO filter by psid -> agent id mappings for multi-page setup
```
[the remark this answers is not in the document]
Comment (JackUrb) [reply parlai/messenger/core/messenger_manager.py]: I'm pretty sure we can remove this, we were investigating having multiple pages talking to the same world to have different roles, but now that's kinda silly as we have other ways to set up multiple roles

File: parlai/messenger/core/shared_utils.py [hunk 2]
```diff
@@ -50,3 +54,118 @@ def print_and_log(level, message, should_print=False):
 def generate_event_id(worker_id):
     """Return a unique id to use for identifying a packet for a worker"""
     return '{}_{}'.format(worker_id, uuid.uuid4())
+
+
+##################
+## New Utils
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/shared_utils.py]: this was more of a bookkeeping thing for me when i was adding new utils; will remove

File: parlai/messenger/core/shared_utils.py [hunk 3]
```diff
@@ -50,3 +54,118 @@ def print_and_log(level, message, should_print=False):
...
+                world_name=task_name,
+                onboarding_name=configuration.get("onboard_world"),
+                task_name=configuration.get("task_world"),
+                max_time_in_pool=configuration.get("timeout") or 300,
+                agents_required=configuration.get("agents_required") or 1,
+                backup_task=configuration.get("backup_task"),
+            )
+        # get world options
+        result["world_opt"] = cfg.get("opt", {})
+        # get local hosting information, if any exists
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/shared_utils.py]: oops this is an artifact, will delete

File: parlai/messenger/core/messenger_manager.py [hunk 4]
```diff
@@ -66,20 +125,18 @@ class MessengerManager:
...
-        self.overworld_threads = {}
         self.active_worlds = {}
         self.message_socket = None
         self.message_sender = None
-        self.page_id = None
         self._init_logs()
         self.running = True
         self.conversation_index = 0
-        self.shutting_down = True
+        self.shutting_down = False
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/messenger_manager.py]: 👀

File: parlai/messenger/core/messenger_manager.py [hunk 5]
```diff
@@ -89,36 +146,93 @@ def __init__(self, opt):
...
+            self._log_debug('Adding agent {} to pool...'.format(agent.messenger_id))
             # time agent entered agent_pool
             agent.time_in_pool.setdefault(world_type, time.time())
             # add agent to pool
             self.agent_pool.setdefault(world_type, []).append(agent)
 
     def mark_removed(self, agent_id, pageid):
-        """Mark the agent as removed from the pool. Can be overriden to change
-        other metadata linked to agent removal."""
+        """Mark the agent as removed from the pool.
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/messenger_manager.py]: not sure, i left in as it was there before?

File: parlai/messenger/core/shared_utils.py [hunk 6]
```diff
@@ -50,3 +54,118 @@ def print_and_log(level, message, should_print=False):
...
+class TaskState:
+    """Wrapper for an agent running on a Worker"""
+
+    def __init__(
+        self, task_name, world_name, agents, is_overworld=False, world_type=None
+    ):
+        self.task_name = task_name
+        self.world_name = world_name
+        self.agents = agents
+        self.is_overworld = is_overworld  # (bool): overworld or task world
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/shared_utils.py]: i'll remove references to state transfer

File: parlai/messenger/core/messenger_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/messenger/core/messenger_manager.py]: i've removed the TODO

**Comment by klshuster** [comment]
i've updated the docs, would appreciate a quick sign-off from either @JackUrb or @emilydinan before merging (also i addressed all the comments)

**Pull request merged**